lib/kafka/connection.rb

```diff
@@ -89,6 +102,12 @@ def request(api_key, request, response_class)
           end
         end
       end
+    rescue Errno::EPIPE, Errno::ECONNRESET, Errno::ETIMEDOUT => e
+      @logger.error "Connection error: #{e}"
+
+      close
```

Comment:
One important design decision is to have a centralized retry loop. Since a typical produce request would include sending message batches to more than one broker, it makes sense to postpone a retry until messages have been sent to the other brokers.